### PR TITLE
Extract snapshotting logic to Snapshotter class

### PR DESCRIPTION
**Pull request by cbracken**
Extract snapshotting logic to Snapshotter class

Extract a Snapshotter class that can be shared between FLX snapshotting,
AOT snapshotting, and assembly AOT snapshotting. Allows for better
testability of snapshotting logic.

* Extracts script snapshotting used in FLX build.
* Adds tests for snapshot checksumming, build invalidation/skipping.

Remaining work: disentangle + extract AOT snapshotting and Assembly AOT
snapshotting logic from build_aot.dart.

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/build.dart [hunk 1]
```diff
@@ -109,3 +152,111 @@ Future<Set<String>> readDepfile(String depfilePath) async {
       .where((String path) => path.isNotEmpty)
       .toSet();
 }
+
+/// Dart snapshot builder.
+///
+/// Builds Dart snapshots in one of three modes:
+///   * Script snapshot: architecture-independent snapshot of a Dart script core
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/base/build.dart]: this seems to switch between singular and plural mid-sentence?

Comment (cbracken) [reply packages/flutter_tools/lib/src/base/build.dart]: Missed a word! Done.

File: packages/flutter_tools/lib/src/commands/build_aot.dart [hunk 2]
```diff
@@ -356,7 +357,7 @@ Future<String> _buildAotSnapshot(
     final Set<String> snapshotInputPaths = await readDepfile(dependencies)
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/commands/build_aot.dart]: (I can't add a comment to the appropriate lines in this file)

Maybe add a todo in this file where gen_snapshot is referenced to indicate that this should be using the new snapshotter instead?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/build_aot.dart (hunk 2 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/commands/build_aot.dart]: Good point! Done.

File: packages/flutter_tools/lib/src/base/build.dart [hunk 3]
```diff
@@ -6,11 +6,54 @@ import 'dart:async';
...
+    final String isolateSnapshotData = artifacts.getArtifactPath(Artifact.isolateSnapshotData);
+    final List<String> args = <String>[
+      '--assert_initializer',
+      '--await_is_keyword',
+      '--causal_async_stacks',
+      '--vm_snapshot_data=$vmSnapshotData',
+      '--isolate_snapshot_data=$isolateSnapshotData',
+      '--packages=$packagesPath',
+      '--dependencies=$depfilePath',
+      '--print_snapshot_sizes',
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/build.dart]: can we remove this --print option? It adds spurious output when running `flutter run --release`.

Comment (Hixie) [reply packages/flutter_tools/lib/src/base/build.dart]: https://github.com/flutter/flutter/pull/18049

**Pull request merged**